app/services/migration/migrators/base.rb

```diff
@@ -83,8 +83,48 @@ def data_migration
       @data_migration ||= Migration::DataMigration.find_by(model: self.class.model, worker:)
     end
 
+    def find_lead_provider_id!(ecf_id:)
+      lead_provider_ids_by_ecf_id[ecf_id] || raise(ActiveRecord::RecordNotFound, "Couldn't find LeadProvider")
+    end
+
+    def find_cohort_id!(start_year:)
+      cohort_ids_by_start_year[start_year] || raise(ActiveRecord::RecordNotFound, "Couldn't find Cohort")
+    end
+
+    def find_application_id!(ecf_id:)
+      application_ids_by_ecf_id[ecf_id] || raise(ActiveRecord::RecordNotFound, "Couldn't find Application")
+    end
+
+    def find_declaration_id!(ecf_id:)
+      declaration_ids_by_ecf_id[ecf_id] || raise(ActiveRecord::RecordNotFound, "Couldn't find Declaration")
+    end
+
+    def find_statement_id!(ecf_id:)
+      statement_ids_by_ecf_id[ecf_id] || raise(ActiveRecord::RecordNotFound, "Couldn't find Statement")
+    end
+
   private
 
+    def statement_ids_by_ecf_id
+      @statement_ids_by_ecf_id ||= ::Statement.pluck(:ecf_id, :id).to_h
+    end
+
+    def declaration_ids_by_ecf_id
+      @declaration_ids_by_ecf_id ||= ::Declaration.pluck(:ecf_id, :id).to_h
+    end
+
+    def application_ids_by_ecf_id
+      @application_ids_by_ecf_id ||= ::Application.pluck(:ecf_id, :id).to_h
+    end
+
+    def lead_provider_ids_by_ecf_id
+      @lead_provider_ids_by_ecf_id ||= ::LeadProvider.pluck(:ecf_id, :id).to_h
+    end
+
+    def cohort_ids_by_start_year
+      @cohort_ids_by_start_year ||= ::Cohort.pluck(:start_year, :id).to_h
+    end
```

Comment:
Should we use redis cache or memcache for this? that way all the workers will shared the same cache.
Also, I wonder if this is using up all the RAM, redis would resolve that.

Reply:
I actually tried this as part of the investigation but our Redis instance is quite small and it ended up being a bottleneck. I think there was also contention around which worker populates the cache (as you get ~20 trying to do it at the same time). I have the changes stashed if we want to revisit it/bump our Redis instance though.